src/reducers/fileTree.tsx

```diff
@@ -157,6 +159,74 @@ export const buildFileTree = (version: Version): FileTree => {
   };
 };
 
+type GetRelativePathParams = {
+  _log: typeof log;
+  currentPath: string;
+  pathList: string[];
+  position: 'previous' | 'next';
+};
+
+export const getRelativePath = ({
+  _log = log,
+  currentPath,
+  pathList,
+  position,
+}: GetRelativePathParams): string | void => {
+  const currentIndex = pathList.indexOf(currentPath);
+  if (!currentIndex) {
+    _log.debug(`Cannot find ${currentPath} in pathList: ${pathList}`);
+    return undefined;
+  }
+  const newIndex =
+    position === 'previous' ? currentIndex - 1 : currentIndex + 1;
+  const validIndex = newIndex >= 0 && pathList.length >= newIndex;
+  if (!validIndex) {
+    _log.debug(`There is not a ${position} file wrt ${currentPath}`);
+  }
+  return validIndex ? pathList[newIndex] : undefined;
+};
+
+type GoToNextFileParams = {
+  _log: typeof log;
+  currentPath: string;
+};
+
+export const goToNextFile = ({
+  _log = log,
+  currentPath,
+}: GoToNextFileParams): ThunkActionCreator => {
+  return async (dispatch, getState) => {
+    const { fileTree: fileTreeState } = getState();
+    const { forVersionId, tree } = fileTreeState;
+    const { pathList } = tree;
+
+    if (!forVersionId || !pathList) {
+      _log.debug('Cannot navigate to next file with no version loaded');
```

Comment:
Since this thunk will be triggered from a key press, it seems likely that the user will encounter this case. You could dispatch an action here to show a message that the data is still loading, similar to my other comment.

I wonder if it's better to read these variables from action parameters rather than from state but I'm not sure. If you read them from parameters, the controlling component would know when to load data (when they are empty) but maybe that's not necessary. You could keep it reading from state (as is) for now.

Reply:
@kumar303 Is your suggestion here that the component dispatching `gotToNextFile` could pass the state variables into the action? I can see how that could be useful for your first comment, as we wouldn't even bother dispatching unless a version was loaded. In fact, the keyboard shortcuts don't make any sense without a tree, so we could not even render the shortcut list until the tree is present. This makes a lot of sense, but I'm not sure passing that amount of data is a good idea. The action needs the entire `pathList` from the tree, which could be quite large. Is that considered okay?